prov/gni/src/gnix_msg.c

```diff
@@ -55,6 +55,94 @@
 /*******************************************************************************
  * helper functions
  ******************************************************************************/
+static void __gnix_msg_unpack_data_into_iov(const struct recv_info_t *dest,
+					    size_t dest_cnt,
+					    uint64_t src, size_t src_len)
+{
+	static int i;
+	static size_t cum_len;
+	static size_t offset;
+	size_t bwrite;
+
+	GNIX_DEBUG(FI_LOG_EP_DATA, "Unpacking data for recvv count (%lu)"
+		   " src = %p, dest = %p,"
+		   " i = %d, cum_len = %lu, offset = %lu\n",
+		   dest_cnt, src, dest, i, cum_len, offset);
+
+	/* Pull out each iov len and base field into the iov */
+	while (i < dest_cnt) {
+		bwrite = dest[i].recv_len - offset < src_len - cum_len ?
+			dest[i].recv_len - offset : src_len - cum_len;
+
+		memcpy((uint8_t *) dest[i].recv_addr + offset,
+		       (void *) (src + cum_len), bwrite);
+
+		cum_len += bwrite;
+
+		/* Src exhausted */
+		if (cum_len == src_len) {
+			offset = dest[i].recv_len - bwrite;
+			cum_len = 0;
+
+			/* dest exhausted too, reset the counters */
+			if (!offset) {
+				goto fini;
+			}
+
+			return;
+		}
+
+		i++;
+	}
+
+fini:
+	/* dest exhausted, reset the counters */
+	i = cum_len = 0;
+}
+
+static void __gnix_msg_pack_data_from_iov(uint64_t dest, size_t dest_len,
+					  uint64_t src, size_t src_cnt)
```

Comment:
> Why do you pass a vector to this function as "uint64_t src"?

To avoid casting and dereferencing the iov for performing pointer arithmetic and then casting the iov back to a void pointer for memcpy.